src/inc/til/u8u16convert.h

```diff
@@ -306,25 +91,160 @@ namespace til // Terminal Implementation Library. Also: "Today I Learned"
+                IsLead3   = gsl::narrow_cast<char>(0b1110'0000), // A lead byte that indicates a UTF-8 non-ASCII character consisting of three bytes has the three highest bits set to 1 and the adjacent bit set to 0
+                MaskLead3 = gsl::narrow_cast<char>(0b1111'0000), // Bit mask to be used in a bitwise AND operation to find out whether or not a lead byte match the IsLead3 pattern
+                IsLead4   = gsl::narrow_cast<char>(0b11110'000), // A lead byte that indicates a UTF-8 non-ASCII character consisting of four bytes has the four highest bits set to 1 and the adjacent bit set to 0
+                MaskLead4 = gsl::narrow_cast<char>(0b11111'000), // Bit mask to be used in a bitwise AND operation to find out whether or not a lead byte match the IsLead4 pattern
+            };
+            /* clang-format on */
+
+            constexpr const std::array<char, 4> cmpMasks{
+                '\0', // unused
+                MaskCont,
+                MaskLead2,
+                MaskLead3,
+            };
+
+            constexpr const std::array<char, 4> cmpOperands{
+                '\0', // unused
+                IsAscii, // intentionally conflicts with MaskCont
+                IsLead2,
+                IsLead3,
+            };
+
+            int capa16{};
+            // The worst ratio of UTF-8 code units to UTF-16 code units is 1 to 1 if UTF-8 consists of ASCII only.
+            RETURN_HR_IF(E_ABORT, !base::CheckAdd(in.length(), state.partials[3]).AssignIfValid(&capa16));
+
+            out.clear();
+            RETURN_HR_IF(S_OK, !capa16); // nothing to convert, however this is not an error
+
+            out.resize(gsl::narrow_cast<size_t>(capa16));
+            auto len8{ gsl::narrow_cast<int>(in.length()) };
+            int len16{};
+            auto cursor8{ in.data() };
+            if (!!state.partials[3])
+            {
+                if (!len8) // only the partial code point will be converted and published
+                {
+                    len16 = MultiByteToWideChar(CP_UTF8, 0UL, state.partials.data(), gsl::narrow_cast<int>(state.partials[3]), out.data(), capa16);
+                    RETURN_HR_IF(E_UNEXPECTED, !len16);
+
+                    state.reset();
+                    out.resize(gsl::narrow_cast<size_t>(len16));
+                    return S_FALSE;
+                }
+
+                // determine the number of bytes of the code point indicated by the lead byte in the cache
+                const auto codePointLen{
+                    2 + // the code point consists of at least 2 bytes because if we cached data it will never be ASCII
+                    gsl::narrow_cast<int>((state.partials[0] & IsLead3) == IsLead3) + // add one if 3 high-order bits of the lead byte are set
+                    gsl::narrow_cast<int>((state.partials[0] & IsLead4) == IsLead4) // add another one if 4 high-order bits of the lead byte are set
+                };

```

Comment:
If you store the number of remaining/missing characters in `state.partials[3]` instead, you don't need to calculate this.

Reply:
I have to think about this. My current understanding is that we would have to evaluate the lead byte anyways. How would you know the length of the sequence? It's not the same as number of remaining bytes in the cache ¯\_(ツ)_/¯

Reply:
Oh right because you want to pass the length of the input buffer to `MultiByteToWideChar`.
In that case I'd probably define:
```cpp
struct u8state {
	char buffer[4];
	uint8_t want = 0;
	uint8_t have = 0;
};
```
...or something similar. It would still allow us to get rid of the "redundant" cmpMasks. I don't think we have to be _that_ mindful about the size of the structure, since the current solution is 44 bytes large (and it's only used in 3 places).

Reply:
Yea that sounds quite reasonable. FWIW I'm not worrying about 2 bytes more. We still shrink it and that's nothing compared to the buffer we allocated 😄